### PR TITLE
Improve yield from inference for unions of generators

### LINKED ISSUE #15141: Starting from version 1.0.0, a union of Generators is (sometimes) treated as if it has no ReturnType
**Bug Report**

Starting from version 1.0.0, `yield from` a union on `Generator` types generates the error `Function does not return a value  [func-returns-value]`, and attempting to use the return reveals that Mypy is, in fact, treating it as if it's None.

There are ways around this; it appears that the inference is a little smarter at function boundaries than in-line, so I've just defined a bunch of helper functions to work around it for now.

**To Reproduce**

```python
from typing import Generator, TypeVar

T = TypeVar("T")


def my_iter(arg: Generator[int, None, T] | Generator[str, None, T]) -> Generator[int | str, None, T]:
    return (yield from arg)
```
(https://mypy-play.net/?mypy=0.991&python=3.11&gist=5c3183a0d4d1c02161211b01654d519f)

(The repro is working with `YieldType`, but in my code it's changing `ReturnType`. It breaks either way around.)

**Actual Behavior**

On post-1.0.0 versions:
```
main.py:7: error: Function does not return a value  [func-returns-value]
main.py:7: error: Incompatible return value type (got "None", expected "T")  [return-value]
Found 2 errors in 1 file (checked 1 source file)
```
Pre-1.0.0, produces no output.

### PR DESCRIPTION
**Pull request by hauntsaninja**
Fixes #15141, closes #15168

### CODE REVIEW

**Comment by hauntsaninja** [comment]
New error in rich is a true positive

File: mypy/checker.py [hunk 1]
```diff
@@ -992,8 +992,9 @@ def get_generator_return_type(self, return_type: Type, is_coroutine: bool) -> Ty
             # AwaitableGenerator, Generator: tr is args[2].
             return return_type.args[2]
         else:
-            # Supertype of Generator (Iterator, Iterable, object): tr is any.
-            return AnyType(TypeOfAny.special_form)
+            # We have a supertype of Generator (Iterator, Iterable, object)
+            # Treat `Iterator[X]` as a shorthand for `Generator[X, Any, None]`.
```
Comment (JelleZijlstra) [inline mypy/checker.py:996]: Why? Intuitively Any makes more sense to me here.

This affects code like this, right?
```python
x: Iterator[T]
y = yield from x
```

File: test-data/unit/check-statements.test [hunk 2]
```diff
@@ -2231,6 +2231,32 @@ class B: pass
...
+
+def foo(arg: Union[T, R]) -> Generator[Union[int, str], None, T]:
+    return (yield from arg) # E: "yield from" can't be applied to "Union[T, R]"
+
+[case testYieldFromUnionOfGeneratorWithIterableStr]
+from typing import Generator, Union, Iterable, Optional
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Iterable[str]]) -> Generator[Union[int, str], None, Optional[T]]:
```
Comment (JelleZijlstra) [inline test-data/unit/check-statements.test:2256]: This is unsound and should arguably error. For example, I could pass a `Generator[str, None, str]` and get a different type out.

Possibly we should allow it for pragmatic reasons, but then again I'm not sure how often people use `yield from` with values typed as just `Iterable`.

File: mypy/checker.py (hunk 1 above)
Comment (hauntsaninja) [reply mypy/checker.py:996]: This results in errors for cases like https://github.com/python/typeshed/issues/11222 / `testReturnInIterator`

File: test-data/unit/check-statements.test (hunk 2 above)
Comment (hauntsaninja) [reply test-data/unit/check-statements.test]: Oh this is a great point. I've added a test, but I think this is basically pre-existing. mypy's existing behaviour is to error with `Function does not return a value (it only ever returns None)  [func-returns-value]` on the return statement

**Pull request merged**